dask_ml/decomposition.py

```diff
@@ -0,0 +1,362 @@
+"""
+Matrix decomposition methods for dimensionality reduction.
+"""
+import dask.array as da
+import numpy as np
+from dask import compute
+from sklearn.base import BaseEstimator, TransformerMixin
+from sklearn.utils.validation import check_is_fitted
+from sklearn.utils.extmath import stable_cumsum
+
+from dask_ml.utils import svd_flip
+
+
+__all__ = [
+    "TruncatedSVD",
+    "PCA"
+]
+
+
+class PCA(BaseEstimator, TransformerMixin):
+    """
+
+    Notes
+    -----
+    Differences from scikit-learn:
+
+    * svd_solver : 'randomized' uses ``dask.linalg.svd_compressed``
+      and 'full' uses ``dask.linalg.svd``. 'arpack' is not valid.
+    * iterated_power : defaults to ``0``, the default for
+      ``dask.linalg.svd_compressed``.
+    * n_components : ``n_components='mle'`` is not allowed
+    """
+
+    def __init__(self, n_components=None, copy=True, whiten=False,
+                 svd_solver="auto", tol=0.0, iterated_power='auto',
+                 random_state=None):
+        self.n_components = n_components
+        self.copy = copy
+        self.whiten = whiten
+        self.svd_solver = svd_solver
+        self.tol = tol
+        self.iterated_power = iterated_power
+        self.random_state = random_state
+
+    def fit(self, X, y=None):
+        self._fit(X)
+        return self
+
+    def _fit(self, X):
+        solvers = {"full", "auto", "tsqr", "randomized"}
+        solver = self.svd_solver
+
+        if solver not in solvers:
+            raise ValueError("Invalid solver '{}'. Must be one of {}".format(
+                solver, solvers
+            ))
+
+        if solver == 'auto':
+            solver = 'full'
+
+        # Handle n_components==None
+        if self.n_components is None:
+            # TODO: handle nan shapes
+            n_components = min(X.shape)
+        else:
+            n_components = self.n_components
+
+        if solver == 'full':
+            return self._fit_full(X, n_components)
+        else:
+            return self._fit_truncated(X, n_components)
+
+    def _fit_full(self, X, n_components):
+        # TODO: unknown shapes
+        # TODO: dataframe input
+        n_samples, n_features = X.shape
+        self.mean_ = X.mean(0)
+        X -= self.mean_
+
+        U, S, V = da.linalg.svd(X)
+        U, V = svd_flip(U, V)
+
+        explained_variance_ = (S ** 2) / (n_samples - 1)
+        total_var = explained_variance_.sum()
+        explained_variance_ratio_ = explained_variance_ / total_var
+
+        # compute
+        (U, S, V,
+         explained_variance_, total_var,
+         explained_variance_ratio_,
+         ) = compute(U, S, V,
+                     explained_variance_, total_var,
+                     explained_variance_ratio_)
+        components_ = V
+        singular_values_ = S.copy()
```

Comment:
Is the compute call above safe?  Are the U and V parts of that computation likely to be large?

Do we ever use the computed `U` matrix explicitly?

Reply:
- `components_` (`V`)  is `(min(n_components, n_features_), n_features)`
- `S` and `U` are both `(min(n_components, n_features)`,)

So as long as the array is not too wide, this will be OK.  I've mainly been assuming this up till now, though decomposition is probably the area we're mostly likely to run into extremely wide arrays.